### ISSUE TITLE
DreamerV2のActorモデルのバグ

### ISSUE THREAD

**Issue by seki-shu** [seki-shu/Group15_Multimodal#5]
#3  @pchan1111 

以下のコードの `action_log_probs = action_dist.log_prob(unscaled_actions) - torch.log(1 - actions.pow(2) + 1e-7).sum(dim=1, keepdim=True)` に関して， ` action_log_probs : torch.Tensor (batch size, 1) 最適行動の対数尤度.` となっていますが，実際の形が`(batch size, batch size)`になっていて，バグになっています．修正案の検討をお願いします．
```

class Actor(nn.Module):
    """
    最適行動を出力するモデル.
    今回はAtariではないので連続な行動を出力する.

    Methods:
    -------
        forward: 最適行動を出力する.
    """
    def __init__(self, action_dim: int, h_dim: int, z_dim: int, n_classes: int, hidden_dim: int):
        """
        コンストラクタ.

        Params:
        -------
            action_dim : int
                行動の次元数.
            h_dim : int
                決定的状態hの次元数.
            z_dim : int
                確率的状態zの次元数.
            n_classes : int
                確率的状態zのカテゴリ数.
            hidden_dim : int
                MLPを通した後のユニット数.
        """
        super().__init__()
        self.fc1 = nn.Linear(h_dim + z_dim * n_classes, hidden_dim)
        self.fc2 = nn.Linear(hidden_dim, hidden_dim)
        self.fc3 = nn.Linear(hidden_dim, hidden_dim)
        self.fc4 = nn.Linear(hidden_dim, hidden_dim)
        self.fc_mean = nn.Linear(hidden_dim, action_dim)
        self.fc_std = nn.Linear(hidden_dim, action_dim)

    def forward(self, h: torch.Tensor, z: torch.Tensor, eval: bool):
        """
        最適行動を出力する.
        a_t = f(h_t, z_t)

        Params:
        -------
            h : torch.Tensor (batch size, h_dim)
                決定的状態h.
            z : torch.Tensor (batch size, z_dim * n_classes)
                確率的状態z.
            eval : bool
                評価モードかどうか.

        Returns:
        -------
            actions : torch.Tensor (batch size, action_dim)
                最適行動.
            action_log_probs : torch.Tensor (batch size, 1)
                最適行動の対数尤度.
            action_entropy : torch.Tensor (batch size, 1)
                最適行動のエントロピー.
        """
        x = F.elu(self.fc1(torch.cat([h, z], dim=1)))
        x = F.elu(self.fc2(x))
        x = F.elu(self.fc3(x))
        x = F.elu(self.fc4(x))
        mean = self.fc_mean(x) # (batch_size, action_dim)
        std = F.softplus(self.fc_std(x))

        if eval:
            actions = torch.tanh(mean) # action spaceを(-1, 1)にしてる. 環境によっては変更必要かも.
            return actions, None, None

        # Dreamer V2の連続空間でのActorがどうなっているかいまいち分からなかったので、とりあえず以下のように実装しました。
        # まずactionを正規分布からサンプルした後、tanh関数に通して(-1, 1)にします。
        # 対数尤度を求めるときに、log p(actions) = log Normal(unscaled_actions) * |det(∂unscaled_actions / ∂actions)|(https://arxiv.org/pdf/1801.01290 Appendix C)となります
        # tanhの微分は1-tanh^2で、正規分布の分散共分散行列は対角行列なので、ヤコビアンはdiag(1 - tanh^2)となります。
        # 間違ってるかもしれません...
        action_dist = td.Independent(Normal(mean, std), 1)
        unscaled_actions = action_dist.rsample()
        actions = torch.tanh(unscaled_actions)

        action_log_probs = action_dist.log_prob(unscaled_actions) - torch.log(1 - actions.pow(2) + 1e-7).sum(dim=1, keepdim=True)
        action_entropy = action_dist.entropy()
        return actions, action_log_probs, action_entropy
```

**Comment by pchan1111** [comment]
action_dist.log_prob(unscaled_actions)
の形が(B, 1)ではなく(B, )となりバグが発生していたので、
unsqueeze(-1)
を追加して(B, 1)に形を変更してバグを修正しました！
